src/cache.ts

```diff
@@ -0,0 +1,82 @@
+import crypto from "node:crypto";
+import fs from "node:fs/promises";
+import path from "node:path";
+import { mkdirSync } from "node:fs";
+
+export class Cache {
+  private dir: string;
+  private loading: Record<string, Promise<unknown>>;
+  private isDisabled: boolean;
+
+  constructor(dir: string, isDisabled = false) {
+    this.dir = dir;
+    this.loading = {};
+    this.isDisabled = isDisabled;
+
+    if (!this.isDisabled) {
+      mkdirSync(this.dir, { recursive: true });
+    }
+  }
+
+  private key(key: string) {
+    return crypto.createHash("sha256").update(key).digest("hex");
+  }
+
+  private filename(key: string) {
+    return path.join(this.dir, this.key(key));
```

Comment:
As someone reading the code, it's a little confusing why a `key` function would be needed here since the key has already been provided. Changing the name slightly for either the function or the parameter might clarify that, e.g.:

```
private filepath(key: string) {
  return path.join(this.dir, this.filenameForKey(key))
}
```